internal/parentpath/has_list_or_set.go

```diff
@@ -0,0 +1,20 @@
+package parentpath
+
+import "github.com/hashicorp/terraform-plugin-framework/path"
+
+// HasListOrSet returns true if any step of given path is a list or set. This
+// cannot detect if the last step is a list or set.
```

Comment:
For my own curiosity, what are the scenarios where the last path step would be a list or set? Empty?

Reply:
The path information is inherently lossy in terms of what the last step of the path represents, because each path step is intended to provide the "next step" into the data from the previous step. (If you ever see `ApplyTerraform5AttributePathStep` methods, thats essentially this stuff.)

The easiest example of the last step being undetectable as a list/set is the below, where the path implementation is  `path.Path{steps: path.PathSteps{PathStepAttributeName("list")}}`, where the absolute root of a schema path is an object of attributes, only traversable via attribute name steps.

```go
schema.Schema{
  Attributes: map[string]schema.Attribute{
    "list": schema.ListAttribute{
       ElementType: types.StringType,
       PlanModifiers: []planmodifier.List{
         // planmodifier.ListRequest.Path will be path.Root("list") for each of these
         // The last step of the path doesn't know that the attribute is a list.
       },
     },
  },
}
```

Reply:
Oh and just in case its not clear, the way to detect what the last step of a path is can be done by walking the entire path to fetch the data, then using whatever data implementation details to determine what that data actually is (e.g. type checking against interfaces or implementation types).

If we needed to do this within an attribute plan modifier implementation, the request plan value would already contain the value for type checking, since the framework already does all the work to fetch the data at the path as part of creating the requests.